### PR TITLE
[multipart] video previews and content progress page

### PR DESCRIPTION
**Pull request by schatten**
Summary
---------
A lot of the work was touching the same files, so rather than split into multiple PRs, here is one to rule them all.

### [`291502c3`](https://github.com/facebook/ThreatExchange/pull/763/commits/291502c3fd04b51f9aa040f355a52c23b03d59d5) Add a single API for progress of a content through the pipeline

Combines DynamoDB queries across ActionEvents, MatchRecords, HashRecords, ContentObjects and returns a single view for the progress of the content through the pipeline.

### [`48c73bd7`](https://github.com/facebook/ThreatExchange/pull/763/commits/48c73bd78b8ea850b4ba3a104db8412fbfacb67d) Use CSS for image blurs, allow external governing of blur state

This allows for behavior other than unblur-on-hover. Used later to power click-to-unblur behavior.

### [`daf7e067`](https://github.com/facebook/ThreatExchange/pull/763/commits/daf7e067fb4a94c8cb7c13b9ead066ae9c4b07ab) [breaking] Migrate TextFieldsUtils to typescript (click-to-copy)

I was unable to accurately produce the types required for UpdatablePopover. If this is essential (might-be), we'll need to dive deeper into react and react-bootstrap type defs. Those are not easily grokkable.

Estimate an hour of work to \[re\]fix this. Apologies.

### [`947de7f5`](https://github.com/facebook/ThreatExchange/pull/763/commits/947de7f5f07c0354088e5bbd42425d016c5170f9) Add a page to show stages that content goes through with their times

This does not yet include the actual hash values and the matching datasets.

### [`d769bf7c`](https://github.com/facebook/ThreatExchange/pull/763/commits/d769bf7c2cfa6084b0f5726a0f7982ed59ce118b) PipelineProgress: Poll until action_performed_at is defined

Refetches the pipeline progress every 5 seconds. 

### [`4f29eddb`](https://github.com/facebook/ThreatExchange/pull/763/commits/4f29eddb7751583fbf5a0bba4360675a76f9ef55) Use signed URLs instead of sending down raw bytes in /image/; circumvent CORS

With the earlier image bytes, CDN urls broke. CDN URLs can't be fetched in an AJAX call because of CORS. Instead, I'm using a signed URL for direct s3 uploads and the original content_ref URL for via-URL uploads. Using native &lt;img src&gt; style tags to render images and videos.

# Ignorable Typescript migrations

### [`d6ffce1e`](https://github.com/facebook/ThreatExchange/pull/763/commits/d6ffce1e3a5d2b1ab268dd4eb3c64e10b42e7292) Relax eslint rules which prevent usage of typescript enums

My autoformatter attempted to "fix" the entire file. The only interesting bit is adding of a rule ~L83.

Test Plan
---------
Uploaded a GIF using S3 and separately uploaded a video and an image using URLs.

Combined screenshot of content-details page is attached.
Screen Shot 2021-08-18 at 16 59 51

You'll notice that the hover to unblur previews were replaced with a more explicit "click-button-to-reveal" previews. I'm trying this out because I feel explicit is better here, but open to changing it and it is a oneline change.

The webapp was built and deployed. And the progress page loaded by manually entering the URL. `/pipeline-progress/<content_id>/`.

Here's an incomplete pipeline run.
Screen Shot 2021-08-18 at 17 04 08

Here's a complete one.
Screen Shot 2021-08-18 at 17 04 48

### CODE REVIEW

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (BarrettOlson) [inline hasher-matcher-actioner/hmalib/lambdas/api/content.py:42]: nit: if it works for video why call it `ImageResponse` ?

File: hasher-matcher-actioner/webapp/src/pages/ContentDetails.jsx
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/pages/ContentDetails.jsx:41]: rm

File: hasher-matcher-actioner/webapp/src/pages/ContentDetailsWithStepper.tsx
Comment (BarrettOlson) [inline hasher-matcher-actioner/webapp/src/pages/ContentDetailsWithStepper.tsx:93]: aside: curious, what does the `!` do?

**Review by BarrettOlson — COMMENTED** [review]
Reviewed commit-by-commit:

> 291502c3 Add a single API for progress of a content through the pipeline
> Combines DynamoDB queries across ActionEvents, MatchRecords, HashRecords, ContentObjects and returns a single view for the progress of the content through the pipeline.

LGTM, not sure about selecting min time vs max (comments in line) 


> 48c73bd7 Use CSS for image blurs, allow external governing of blur state
> This allows for behavior other than unblur-on-hover. Used later to power click-to-unblur behavior.

LGTM

> daf7e067 [breaking] Migrate TextFieldsUtils to typescript (click-to-copy)
> I was unable to accurately produce the types required for UpdatablePopover. If this is essential (might-be), we'll need to dive deeper into react and react-bootstrap type defs. Those are not easily grokkable.
> Estimate an hour of work to [re]fix this. Apologies.

Sure. (for now I'll take the loss tiny feature for the addition of typing)

> 947de7f5 Add a page to show stages that content goes through with their times
This does not yet include the actual hash values and the matching datasets.

LGTM, the stage pieces of progress-stepper feel like they could be made into methods for easier maintenance and reuse (probably wouldn't reduce line count right now though) 

> d769bf7c PipelineProgress: Poll until action_performed_at is defined
Refetches the pipeline progress every 5 seconds.

LGTM, nit/aside: thoughts on having auto/poll/refresh be configurable? (a button that sets/resets `pollBuster`) or even have refresh timeout backoff if nothing changes?

> dcfd564b Make ContentPreview generic: now does video previews

Nice! would love to see and example in the test plan.

> 12d3f038 Render video and photos in ContentDetailsWithStepper

LGTM.

> 4f29eddb Use signed URLs instead of sending down raw bytes in /image/; circumvent CORS
With the earlier image bytes, CDN urls broke. CDN URLs can't be fetched in an AJAX call because of CORS. Instead, I'm using a signed URL for direct s3 uploads and the original content_ref URL for via-URL uploads. Using native  style tags to render images and videos.

Nice fix!

File: hasher-matcher-actioner/webapp/src/pages/ContentDetails.jsx
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/pages/ContentDetails.jsx:41]: surprised the linter didn't complain, we should probably make the full change over to .tsx as I think that is likely the reason.

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:42]: (if it is because of the path of the endpoint we can just updated that as well yeah?)

**Review by BarrettOlson — APPROVED** [review]
Approving to unblock, two considerations I'd like made before merge is  1) the `min` vs `max` question made inline and 2) an addition to the UI either directly or in a tool tip that clearly informs the user that not all content should expect a match or action. So it is likely/common to stay at the hashed state indefinitely.  

These changes are cool but also likely to introduce a bit more confusions without additional  explanations of the details (ideally on the page itself). Thanks!

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:41]: Yeah. Have been thinking about this. My usage of models stems from using [django models](https://docs.djangoproject.com/en/3.2/topics/db/models/) which are exclusively for data storage. Messages and HTTP responses are better classified as `representations`. 

But until we socialize this and "make" this a convention, I've resorted to keeping the representation classes close to the APIs so they are not interpreted as models.
for the

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:42]: Yeah. That makes more sense. I was being lazy. I'll make the change. Thanks!

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:148]: That's correct. I'll probably re-use the python function behind the erstwhile /image/ API to get this. Thanks!

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:163]: One of the goals for the pipeline progress indicator was to show the low latency. It actually surprised me too. :) 

Which is why the `min` made sense to me. Also, when you say "most recently..." you mean with the max() value, right? With the current 'min', it would be "earliest ...".

File: hasher-matcher-actioner/hmalib/lambdas/api/content.py
Comment (schatten) [reply hasher-matcher-actioner/hmalib/lambdas/api/content.py:166]: granted. I'll fail loudly instead.

File: hasher-matcher-actioner/webapp/src/components/ContentPreview.tsx
Comment (schatten) [reply hasher-matcher-actioner/webapp/src/components/ContentPreview.tsx:63]: aside-to-your-asied: &nbsp; is HTML. 😆

File: hasher-matcher-actioner/webapp/src/components/ContentProgressStepper.tsx
Comment (schatten) [reply hasher-matcher-actioner/webapp/src/components/ContentProgressStepper.tsx:212]: I did. But I do not see the stepper segments being dynamic. If that happens, then yes. Otherwise, either way solves the same problem.

File: hasher-matcher-actioner/webapp/src/pages/ContentDetailsWithStepper.tsx
Comment (schatten) [reply hasher-matcher-actioner/webapp/src/pages/ContentDetailsWithStepper.tsx:93]: type casts optional<T> to T. Also, will throw a runtime error if contentType is undefined. Hence the check in line 90.

File: hasher-matcher-actioner/webapp/src/utils/TextFieldsUtils.tsx
Comment (schatten) [reply hasher-matcher-actioner/webapp/src/utils/TextFieldsUtils.tsx:51]: It still does that, although looks weird. See the notch when the text changes. 
e8036437898f6b85f2b25d6452b0213d

Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/utils/TextFieldsUtils.tsx:51]: yeah, I remember that bug being annoying to fix. Good enough for me as is

**Pull request merged**
